n3fit/src/n3fit/backends/keras_backend/multi_dense.py

```diff
@@ -0,0 +1,191 @@
+    >>>         ])
+    >>>     for r in range(replicas)
+    >>>     ]
+    >>> gridsize, features = 100, 2
+    >>> multi_dense_model.build(input_shape=(None, gridsize, features))
+    >>> for single_model in single_models:
+    >>>     single_model.build(input_shape=(None, gridsize, features))
+    >>> test_input = tf.random.uniform(shape=(1, gridsize, features))
+    >>> multi_dense_output = multi_dense_model(test_input)
+    >>> single_dense_output = tf.stack([single_model(test_input) for single_model in single_models], axis=1)
+    >>> tf.reduce_all(tf.equal(multi_dense_output, single_dense_output))
+
+    Parameters
+    ----------
+    replica_seeds: List[int]
+        List of seeds per replica for the kernel initializer.
+    kernel_initializer: Initializer
+        Initializer class for the kernel.
+    replica_input: bool (default: True)
+        Whether the input already contains multiple replicas.
+    """
+
+    def __init__(
+        self,
+        replica_seeds: List[int],
+        kernel_initializer: Initializer,
+        replica_input: bool = True,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.replicas = len(replica_seeds)
+        self.replica_seeds = replica_seeds
+        self.kernel_initializer = MultiInitializer(
+            single_initializer=kernel_initializer, replica_seeds=replica_seeds
+        )
+        self.bias_initializer = MultiInitializer(
+            single_initializer=self.bias_initializer, replica_seeds=replica_seeds
+        )
+        self.replica_input = replica_input
+
+        self.matmul = None
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.kernel = self.add_weight(
+            name="kernel",
+            shape=(self.replicas, input_dim, self.units),
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+        )
+        if self.use_bias:
+            self.bias = self.add_weight(
+                name="bias",
+                shape=(self.replicas, 1, self.units),
+                initializer=self.bias_initializer,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+            )
+        else:
+            self.bias = None
+        self.input_spec.axes = {-1: input_dim}
+        self.built = True
+
+        if self.replicas == 1:
+            if self.replica_input:
+                self.matmul = lambda inputs: tf.tensordot(inputs, self.kernel[0], [[-1], [0]])
+            else:
+                # Manually add replica dimension
+                self.matmul = lambda inputs: tf.expand_dims(
+                    tf.tensordot(inputs, self.kernel[0], [[-1], [0]]), axis=1
+                )
+        else:
+            einrule = f"b{'r' if self.replica_input else ''}nf,rfg->brng"
+            self.matmul = lambda inputs: tf.einsum(einrule, inputs, self.kernel)
```

Comment:
Mostly efficiency.

RE the other point, I'd be happy to swap if someone spends the time of reproducing the fits and checking that they are the exactly same. But it needs to be done for all of them.

Reply:
> also we can swap by the end of the year when the theses being written with respect to these 3 papers are finished

This was indeed what I was thinking, if we can remove it from about a year from now we can add a note to say that, by then we can presumably be a bit less strict about the numerical precision with which the current results are reproduced. Unless for efficiency we want to keep it like this anyway.

Reply:
If efficiency is still an issue I'd keep it, but last time I profiled the code was many versions of tensorflow ago.

But indeed, I think in a year we can relax a bit the numerical stability of the results.

Reply:
So we just need a comment linking to this conversation to provide context for the exception of the 1 replica case ;)

Reply:
or... a benchmark

tbh I'm not sure why is it slower, but it is probably also the backpropagation given that's what changes...

Reply:
I thought it was a little slower, but a quick test on my laptop (replacing `if self.replicas == 1` with `False`), on the production runcard I get 0.44s for the tensordot and 0.42s for the einsum (per epoch, average over the second 100 epochs). Anyway since we have to keep it for now, I'll put in the comment, and it'll have to be benchmarked a bit more when it could be removed.